Feature/add yocto light sensor

ThomasMonkman opened the pull request:
Hello! First off this has been a great library to use, thank you very much for making it.

I've added the ability to set the brightness levels based off the ambient light levels from an [yocto light sensor](https://www.yoctopuce.com/EN/products/usb-environmental-sensors/yocto-light-v5-c)

It works exactly like phones or laptops where it will just adjust the light levels based on how bright the room is.

You set a brightness mapping per screen, setting the 0% and 100% brightness at "X" Lux, it then linearly interpolates between that based on sensor readings.

I've been using it for the last few weeks, and it's like automagic :) as I found my self always having to adjust my screens, especially now its getting dark so fast in winter.

It can be toggled per screen, as well light levels set per screen.

### Translations

I've completely left out translations as I wasn't sure how new translations are integrated, I do use a fair amount of information inside strings, so I'm not sure how that would work. But I'm happy to add that in.

I've tired my best to make minimal changes to the repo, and kept in the style you currently have everything.

### Settings

I've also tried to make the settings ui as informative as possible, so hopefully someone can set it up without help

### User Flow

1. install yocto virtual hub service on pc
2. Set password and user name in browser using the virtual hub ui
3. plug in sensor
4. enter connection password and username in the settings
5. adjust brightness of screens to match what you want

### Error handling
I've added error handling as much as I could, here are some screenshots show that off

Everything ok
<img width="1278" height="1377" alt="yocto settings" src="https://github.com/user-attachments/assets/b5af6b3c-9488-43a5-8939-6cb15bce1798" />

No sensor found
<img width="1280" height="1380" alt="yocto sensor not found" src="https://github.com/user-attachments/assets/a040051b-7117-49be-b3dc-dbd317258bf1" />
No connection to virtual hub
<img width="1280" height="1380" alt="yocto virtual hub not found" src="https://github.com/user-attachments/assets/442f4582-67d8-4703-ac76-e866846651a7" />

ThomasMonkman committed:
feat: add brightness based on yocto light sensor

using an yocto light sensor to automatically set the screen brightness based off the ambient light level in lux

it can be toggled per screen, as well light levels set per screen.

ThomasMonkman committed:
tidy: tidy up reconnect logic

xanderfrangos commented:
Just to clarify, I don't expect you to add support for other light sensors. The idea is that if/when support for other light sensors are contributed, they can share the same settings tab/UI.

psxvoid commented:
Hi,

@xanderfrangos Were there any attempts to integrate laptops sensors? Many laptops have ambient light sensors, including mine.

I was thinking that when a laptop has a built-in ambient light sensor and it's connected to an external monitor, it can control it's brightness automatically.

ThomasMonkman commented:
Hi @psxvoid, So while revisiting the pr, I've looked in to the ambient sensors built in to a laptop, as i figured how hard can it be?

What I have found is there is a simple windows UWP api for it.
https://learn.microsoft.com/en-us/uwp/api/windows.devices.sensors.lightsensor?view=winrt-26100

however that is an UWP api, and im not really sure how to consume that in to electron, I've tired [edge-js](https://github.com/agracio/edge-js), but that api isn't avabile that way (although i didn't spend much time on it)

here was the code I quickly tired (just make a bare bones node project with edge-js installed and run this as a script)
```js
var edge = require('edge-js');

var getLight = edge.func({
    source: function() {/*
        using System.Threading.Tasks;
        using Windows.Devices.Sensors;

        public class Startup
        {
            public async Task<object> Invoke(dynamic input)
            {
                var sensor = LightSensor.GetDefault();
                if (sensor == null)
                {
                    return "No light sensor available";
                }

                var reading = sensor.GetCurrentReading();
                return reading.IlluminanceInLux;
            }
        }
    */}
});

getLight(null, function (error, result) {
    if (error) throw error;
    console.log("Light sensor output (lux):", result);
});

```

another option is [nodeRT](https://github.com/NodeRT/NodeRT) but I found that requires installing visual studio 2013, although there are many versions of the package built for different version of windows sdk [here](https://www.npmjs.com/search?q=windows.devices.sensors)
and im not sure how that versioning is handled, can you just use the oldest? or do you have to ship multiple versions? I assume the oldest works forwards. (but then does the ci need vs2013 installed??)

What I am going to do is add in a fake ui sensor, that lets you mimic having a sensor plugged in.

This also lets me test what its like to extend the sensor page with multiple sensors.

I will focus on the changes requested (more generic ui, support for multiple sensors in theory), then maybe look at how to add the ambient light sensor.

ThomasMonkman commented:
@xanderfrangos @psxvoid Ok I've got windows ambient light sensor working in the project using the old COM api you linked above, I need some time to wrap up the ui for it, but I think it shouldn't be much more work.

Then hopefully this should work for a wide range of people :) 

I'm not sure if I should move this pr in to draft? But I need to push some in progress as I'm travelling and don't want it to disappear in to the ether. But it is not ready to review, I will ping when it is.

ThomasMonkman committed:
feat: added support for windows ambient sensor api and start of refactor for multiple sensors

ThomasMonkman committed:
feat: working now, can switch between yocto, windows and a fake light sensor

ThomasMonkman committed:
fix: package lock in sync, and light sensor select styling

ThomasMonkman committed:
wip: remove log change

ThomasMonkman committed:
Merge remote-tracking branch 'upstream/master' into feature/add-yocto-light-sensor

ThomasMonkman commented:
Ok @xanderfrangos @psxvoid I've updated the PR with both a fake Sensor for testing, and a real windows ambient sensor.

Which works on my laptop.

I've done the work in spits and spurts, but I've tired to keep the style consistent.

Typescript would have made this a million times easier, many a curse word was used :P 
I considered adding partial typescript support but figured it would be better as a separate PR.

In the future I think some kind of log, bucket, or fan curve style graph would be better for mapping the Lux to screen brightness.

I did notice a rare bug where the monitors just don't update, feels like some kind of race condition with the monitors field which I get by reference from electron.js, but I didn't have the heart to dig in to the massive electron.js file.

I've kept the monitor mapping field common between all the sensor types, although I don't really like how much text it has, but wanted it to be clear.

Everything is now split out between files and a new sensor should be easy to add.

The window ambient sensor technically supports multiple sensors, although I'm not sure how you would have this on a device.
Currently it just picks the first one, although the UI does show multiple, if they exist, but doesn't allow you to select which one is used.
I just wanted to get this pr over the line, I think selection could be added at a later date, along with not really being able to test it.


![Windows Ambient](https://github.com/user-attachments/assets/a2ed4010-063b-4d3d-923b-735c79a21d6f)

![Fake](https://github.com/user-attachments/assets/e7002663-846e-47b4-9da1-c0c533ec2357)

ThomasMonkman committed:
fix: incorrect varibles left over from refactor

ThomasMonkman committed:
feat: now will heal the light sensor settings, and create a new if not used before

ThomasMonkman commented:
I have just updated to also handle damaged/missing settings file, as well as putting the settings menu back to default...

ThomasMonkman committed:
fix: remove console log

ThomasMonkman committed:
feat: more tidy up

ThomasMonkman committed:
fix: connects now even if disabled, so ui can update with connection status for both windows and yocto

ThomasMonkman commented:
Hello, @xanderfrangos this pr is ready for review. Let me know if you have any questions?

xanderfrangos committed:
Fix missing sensor files from ASAR

xanderfrangos committed:
Fix windows-ambient-sensor not loading from ASAR

xanderfrangos commented:
@ThomasMonkman Sorry for the delay! This looks great from what I've been able to review.

I need to rummage around for a device that might have an ambient light sensor, but the "Fake" option worked as expected in my test. There were a few issues with building the EXE installer that I've resolved. I'll clean up the UI a little, too. You've got all of the pieces there, so it just needs some adjustments.

In the meantime, anyone who wants to can try out this feature using the build artifact here: https://github.com/xanderfrangos/twinkle-tray/actions/runs/23117199584

ThomasMonkman commented:
Cool thank you!

I think most modern laptops should have an ambient light sensor. @psxvoid do you have one? can you test?

@xanderfrangos What changes did you have in mind for the ui? I was never really happy with the description of the light sensor lux mapping.

One issue I've noticed at least on my dell monitor is it stops me be able to access the on screen menu (it closes it very quickly), I'm guessing it closes when ever it receives a command, perhaps it shouldn't change the brightness if there is not real change, although this would have to be based off the screen brightness percentage and not the lux, as lux changes easily.

--

I also think a good guide for setting the mapping is getting your monitors lux/nits output and setting that as the 100%
then setting the minimum to be something like 0% at 5 lux.

here is a lux to nix convertor.
https://calculator.academy/lux-to-nits-calculator/

What you will find is lux goes up crazy fast, a room with almost no lights on is 5 lux
where as outside is somewhere between 20k-100k lux

I'm in medium shade with no direct light on my laptop inside, with only ambient light bouncing off the floor through a window and I'm at 2.2k lux.

With my dell s2421H 5-250 is good as well as my dell u2515h, but I don't have any more modern brighter HDR screens to test.

psxvoid commented on src/light-sensor/sensors/windows-ambient-light-sensor.js:
```diff
@@ -0,0 +1,105 @@
+    this._pollSensors();
+    this.interval = setInterval(() => {
+      this._pollSensors();
+    }, 1000);
```
Maybe make the polling rate configurable via UI? Even for testing it's difficult to say without debugging why lux value is only updated every 5 minutes or so instead of 1 second (on my Yoga 920 laptop in a pretty bad shape).

xanderfrangos committed:
Merge branch 'master' into pr/1125

xanderfrangos committed:
Update package-lock.json

xanderfrangos committed:
Use fake sensor by default to avoid polling

xanderfrangos committed:
Implement sensorPollingInterval setting

xanderfrangos committed:
Update package-lock.json

xanderfrangos committed:
Update UI for light sensor settings

xanderfrangos committed:
Update Light Sensor section description

xanderfrangos committed:
Disconnect when disabling light sensor feature

xanderfrangos committed:
Avoid dividing by 0 or negative values

xanderfrangos committed:
Move new strings to localization system

xanderfrangos committed:
Only apply new brightness if it has changed

xanderfrangos committed:
Fall back to fake sensor for activeSensor

xanderfrangos committed:
Hide light sensor settings when the feature isn't active
